### PR TITLE
[AAP-69065] Idle Timeout for Manager Processes

### PR DESCRIPTION
**Pull request by nickbhasin**
## Description

- What is being changed?
Introduced an idle_timeout configuration option configurable via ansible_platform_manager_idle_timeout host var or gateway_idle_timeout task arg.
- Why is this change needed?
To remove zombie processes which leads to memory consumption.
- How does this change address the issue?
It automatically terminates the manager process and cleans up its socket file after a configured period of inactivity, preventing orphaned processes and resource exhaustion.

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update
- [ ] Test update
- [ ] Refactoring (no functional changes)
- [ ] Development environment change
- [ ] Configuration change

## Self-Review Checklist

- [x] I have performed a self-review of my code
- [ ] I have added relevant comments to complex code sections
- [ ] I have updated documentation where needed
- [ ] I have considered the security impact of these changes
- [ ] I have considered performance implications
- [ ] I have thought about error handling and edge cases
- [ ] I have tested the changes in my local environment

### Steps to Test
1. 
2. 
3. 

### Required Actions

- [ ] Requires documentation updates
  
- [ ] Requires downstream repository changes
  
- [ ] Requires infrastructure/deployment changes
  
- [ ] Requires coordination with other teams
  
- [ ] Blocked by PR/MR: #XXX

### CODE REVIEW

File: plugins/plugin_utils/manager/manager_process.py [hunk 1]
```diff
@@ -26,7 +28,7 @@ def main():
 
     # Read configuration from command line args
     if len(sys.argv) < 10:
-        print(f"ERROR: Expected 9 args, got {len(sys.argv) - 1}", file=sys.stderr)
+        print(f"ERROR: Expected at least 9 args (optional 10th: idle_timeout), got {len(sys.argv) - 1}", file=sys.stderr)
```
Comment (rohitthakur2590) [inline plugins/plugin_utils/manager/manager_process.py:72]: looks like we are writing the sensitive data. (credentials, token) in plaintext on startup and failures.and this is happening in append mode so we are accumulating everything .
Please redact sensitive positions before writing.

Comment (rohitthakur2590) [reply plugins/plugin_utils/manager/manager_process.py:72]: lets also add tests to validate this

File: plugins/plugin_utils/platform/config.py [hunk 2]
```diff
@@ -126,6 +136,10 @@ def extract_gateway_config(task_args: Optional[Dict[str, Any]] = None, host_vars
         request_timeout=gateway_request_timeout,
         connection_mode=connection_mode,
     )
+    if gateway_idle_timeout is not None:
```
Comment (rohitthakur2590) [inline plugins/plugin_utils/platform/config.py]: can we simple include this when we are building dict above

File: tests/unit/plugins/plugin_utils/manager/test_platform_service_idle.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    mock_store.get_auth_credentials.return_value = ("admin", "admin", None)
+    with patch("ansible_collections.ansible.platform.plugins.plugin_utils.manager.platform_manager.get_credential_manager") as mock_cred:
+        mock_cred.return_value.get_or_create_store.return_value = mock_store
+        with patch("ansible_collections.ansible.platform.plugins.plugin_utils.manager.platform_manager._get_requests") as mock_get_requests:
+            mock_get_requests.return_value = mock_requests
+            config = GatewayConfig(base_url="https://127.0.0.1", username="admin", password="admin", idle_timeout=30.0)
+            return PlatformService(config)
+
+
+class TestGatewayConfigIdle(unittest.TestCase):
```
Comment (rohitthakur2590) [inline tests/unit/plugins/plugin_utils/manager/test_platform_service_idle.py:36]: We need to cover token expiration use cases to test idel timeut behaviour

File: plugins/plugin_utils/platform/config.py [hunk 4]
```diff
@@ -104,6 +111,9 @@ def extract_gateway_config(task_args: Optional[Dict[str, Any]] = None, host_vars
         gateway_token = gateway_token_raw
     gateway_validate_certs = task_args.get("gateway_validate_certs") if "gateway_validate_certs" in task_args else host_vars.get("gateway_validate_certs", True)
     gateway_request_timeout = task_args.get("gateway_request_timeout") or host_vars.get("gateway_request_timeout") or 10.0
+    gateway_idle_timeout = task_args.get("gateway_idle_timeout")
```
Comment (rohitthakur2590) [inline plugins/plugin_utils/platform/config.py]: what is this timer ? what is the significance here ?

File: plugins/plugin_utils/manager/manager_process.py [hunk 5]
```diff
@@ -12,6 +12,8 @@
 import traceback
 from pathlib import Path
 
+MANAGER_IDLE_POLL_INTERVAL = int(os.environ.get("ANSIBLE_PLATFORM_IDLE_POLL_SECONDS", "60"))
```
Comment (rohitthakur2590) [inline plugins/plugin_utils/manager/manager_process.py]: IMO instead of having hardcoded ENV var , this should get tuned based on user configured idel timeout value

**Review by rohitthakur2590 — CHANGES_REQUESTED** [review]

File: plugins/plugin_utils/platform/config.py (hunk 4 above)
Comment (nickbhasin) [reply plugins/plugin_utils/platform/config.py]: It sets how long (in seconds) the background manager process can sit idle before shutting itself down, which prevents orphaned zombie processes from accumulating across playbook runs.

File: plugins/plugin_utils/manager/manager_process.py (hunk 1 above)
Comment (nickbhasin) [reply plugins/plugin_utils/manager/manager_process.py:72]: Done

File: tests/unit/plugins/plugin_utils/manager/test_platform_service_idle.py (hunk 3 above)
Comment (nickbhasin) [reply tests/unit/plugins/plugin_utils/manager/test_platform_service_idle.py:36]: Done.

File: plugins/plugin_utils/manager/manager_process.py (hunk 5 above)
Comment (nickbhasin) [reply plugins/plugin_utils/manager/manager_process.py]: Done.

File: plugins/plugin_utils/platform/config.py (hunk 2 above)
Comment (nickbhasin) [reply plugins/plugin_utils/platform/config.py]: Done.

File: plugins/plugin_utils/platform/config.py (hunk 4 above)
Comment (rohitthakur2590) [reply plugins/plugin_utils/platform/config.py]: IMHO the name gateway_idle_timeout sounds misleading as it implies a timeout on the gateway server side (like a session expiry on the gateway), but this actually controls how long the local manager process on the ansible controller stays alive when idle. The gateway never sees this setting. 
we can name it something like persistent_manager_idle_timeout , wdyt ?
also we need to add this timer var in the documentation  in doc_fragments

File: MIGRATION.md [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+Confirm the exact collection version and any Red Hat release notes for your environment; behavior described here follows the current collection implementation.
+
+---
+
+## What changed (summary)
+
+- **Connection plugin `ansible.platform.http`** is the supported way to route Gateway traffic. It implements `get_client()`, which chooses **direct** (ephemeral manager per task) or **persistent** (reuse one manager process and HTTP session across tasks in a play) mode.
+- **Action plugins** for platform modules call `_get_or_spawn_manager()`, which **prefers** the connection plugin's `get_client()` when `ansible_connection` is `ansible.platform.http`. They still work with **`ansible_connection: local`** by spawning an **ephemeral** manager (`spawn_ephemeral_client()`), but that path does **not** integrate with the connection plugin's persistent lifecycle or facts in the same way.
```
Comment (rohitthakur2590) [inline MIGRATION.md]: internal implementation details need not to be provided

Comment (rohitthakur2590) [reply MIGRATION.md]: the whole guide needs to be in docs folder , we should be providing before and after examples, document how users should configure credentials, connection settings, and inventory/playbook parameters.

File: plugins/plugin_utils/platform/config.py [hunk 7]
```diff
@@ -104,6 +123,15 @@ def extract_gateway_config(task_args: Optional[Dict[str, Any]] = None, host_vars
         gateway_token = gateway_token_raw
     gateway_validate_certs = task_args.get("gateway_validate_certs") if "gateway_validate_certs" in task_args else host_vars.get("gateway_validate_certs", True)
     gateway_request_timeout = task_args.get("gateway_request_timeout") or host_vars.get("gateway_request_timeout") or 10.0
+    # How long (seconds) the persistent manager process may sit idle — i.e. receive
+    # no RPC or API traffic — before it shuts itself down and removes its socket.
+    # This prevents orphaned manager processes from accumulating across playbook runs.
+    # Default: 3600 s (1 hour).  Set to 0 to disable idle-based shutdown entirely.
+    # Accepted variable names (task arg takes priority over host var):
+    #   gateway_idle_timeout  /  ansible_platform_manager_idle_timeout
+    gateway_idle_timeout = (
```
Comment (rohitthakur2590) [inline plugins/plugin_utils/platform/config.py]: if gateway_idel_timeout in task vars is 0 and not provided in host vars  , then going through or  chanin here which treats 0 as false and hence will go through to the next source resulting to None
then later on we are setting it to 3600
idle_timeout=(float(gateway_idle_timeout) if gateway_idle_timeout is not None else 3600.0)


this seems like a bug

**Review by rohitthakur2590 — CHANGES_REQUESTED** [review]

File: plugins/plugin_utils/manager/manager_process.py [hunk 8]
```diff
@@ -13,21 +13,63 @@
...
+    Bounds:
+      - Floor: 5 s  — avoids busy-looping for very short timeouts (e.g. tests).
+      - Cap:  60 s  — avoids infrequent checks for very long timeouts.
+      - ``idle_timeout <= 0`` (disabled): returns 60 s (interval is irrelevant).
+
+    The ``ANSIBLE_PLATFORM_IDLE_POLL_SECONDS`` environment variable overrides
+    this calculation entirely and is intended only for test environments where
+    a sub-second or very short poll period is needed.
+    """
+    env_override = os.environ.get("ANSIBLE_PLATFORM_IDLE_POLL_SECONDS")
```
Comment (rohitthakur2590) [inline plugins/plugin_utils/manager/manager_process.py]: for any float val this will crash at the time of return

Comment (rohitthakur2590) [reply plugins/plugin_utils/manager/manager_process.py]: and also this env var looks like exist only so test can shrink the idle timeout , as min time is already 5 seconds . IMO we should avoid having this

File: plugins/plugin_utils/platform/config.py (hunk 4 above)
Comment (nickbhasin) [reply plugins/plugin_utils/platform/config.py]: Done.

File: MIGRATION.md (hunk 6 above)
Comment (nickbhasin) [reply MIGRATION.md]: I have removed this doc from here as discussed earlier. Will raise a new PR for it (include the comments mentioned above).

File: plugins/plugin_utils/manager/manager_process.py (hunk 8 above)
Comment (nickbhasin) [reply plugins/plugin_utils/manager/manager_process.py]: Acknowledged.

File: plugins/plugin_utils/platform/config.py (hunk 7 above)
Comment (nickbhasin) [reply plugins/plugin_utils/platform/config.py]: Yes. I have addressed it.

**Review by rohitthakur2590 — APPROVED** [review]
